kong/developer_service.go

```diff
@@ -0,0 +1,199 @@
+package kong
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+// AbstractDeveloperService handles Developers in Kong.
+type AbstractDeveloperService interface {
+	// Create creates a Developer in Kong.
+	Create(ctx context.Context, developer *Developer) (*Developer, error)
+	// Get fetches a Developer in Kong.
+	Get(ctx context.Context, emailOrID *string) (*Developer, error)
+	// GetByCustomID fetches a Developer in Kong.
+	GetByCustomID(ctx context.Context, customID *string) (*Developer, error)
+	// Update updates a Developer in Kong
+	Update(ctx context.Context, developer *Developer) (*Developer, error)
+	// Delete deletes a Developer in Kong
+	Delete(ctx context.Context, emailOrID *string) error
+	// List fetches a list of Developers in Kong.
+	List(ctx context.Context, opt *ListOpt) ([]*Developer, *ListOpt, error)
+	// ListAll fetches all Developers in Kong.
+	ListAll(ctx context.Context) ([]*Developer, error)
+}
+
+// DeveloperService handles Developers in Kong.
+type DeveloperService service
+
+// Create creates a Developer in Kong.
+// If an ID is specified, it will be used to
+// create a developer in Kong, otherwise an ID
+// is auto-generated.
+func (s *DeveloperService) Create(ctx context.Context,
+	developer *Developer) (*Developer, error) {
+
+	queryPath := "/developers"
+	method := "POST"
+	if developer.ID != nil {
+		queryPath = queryPath + "/" + *developer.ID
+		method = "PUT"
+	}
```

Comment:
PUTs basically just don't behave, even though they're accepted if you only include fields that exist on the developer entity itself, not the attached consumer. POSTs that include IDs do result in a developer with that ID, so the contract is still honored.

```suggestion
// Create creates a Developer in Kong.
// If an ID is specified, it will be used to
// create a developer in Kong, otherwise an ID
// is auto-generated.
// This call does _not_ use a PUT when provided an ID.
// Although /developers accepts PUTs, PUTs do not accept passwords and do not create
// the hidden consumer that backs the developer. Subsequent attempts to use such developers
// result in fatal errors.
func (s *DeveloperService) Create(ctx context.Context,
	developer *Developer) (*Developer, error) {

	queryPath := "/developers"
	method := "POST"
```